### ISSUE TITLE
FPN-21 | Phase 5 | Supervised Retraining Pipeline Using Feedback Data

### ISSUE THREAD

**Issue by LavanyaSrinivas06** [LavanyaSrinivas06/AI-First-Preauth-Fraud-Detection#70]
### Ticket ID: FPN-21
Phase: Phase 5 — Retraining & Continuous Learning
Branch Name: feature/FPN-21-supervised-retraining

---

### Context
Per the thesis proposal, the supervised model (XGBoost) must be periodically retrained with analyst feedback to capture new fraud patterns and reduce false positives.  
This ticket builds a reproducible pipeline that ingests `data/feedback.parquet`, merges with historical data, retrains the model, and versions artifacts.

---

### Scope
- Ingest and validate `data/feedback.parquet`.
- Rebuild training data: combine original splits with feedback rows.
- Retrain XGBoost with consistent preprocessing and SMOTE (train only).
- Version new artifacts and write a retraining report with before/after metrics.

---

### Steps

1. **Data ingestion and merge**
   - Script: `scripts/retrain_supervised.py`
   - Inputs:
     - `data/processed/train.parquet`, `val.parquet`, `test.parquet`
     - `data/feedback.parquet`
     - `artifacts/preprocess.joblib` (unchanged)
   - Actions:
     - Validate feedback schema (see `docs/feedback_schema.md`).
     - Map `analyst_decision` to interim label if no ground truth:
       - `reject` → fraud (1), `approve` → legit (0), `review` → exclude or hold-out (configurable).
     - De-duplicate by `txn_id` (keep most recent).
     - Concatenate labeled feedback into the **train** set only (do not contaminate val/test).
     - Record class balance before/after.

2. **Feature transformation**
   - Apply the existing `preprocess.joblib` to the augmented train set.
   - Apply SMOTE (or SMOTENC) to the transformed training set only.

3. **Model training and tuning**
   - Reuse `scripts/train_xgboost.py` logic or embed within this script.
   - Use a small randomized search for key params (time-bounded).
   - Save metrics on:
     - Validation set (unchanged from original val).
     - Test set (unchanged from original test).
     - Optional: a separate hold-out made from recent feedback if available.

4. **Artifact versioning**
   - Save new model under a timestamped or semver path:
     ```
     artifacts/models/xgb/<YYYYMMDD_HHMM>/xgb_model.pkl
     artifacts/models/xgb/<YYYYMMDD_HHMM>/xgb_metrics.json
     ```
   - Update a lightweight `artifacts/models/xgb/latest.txt` pointer.

5. **Compatibility and rollback**
   - Verify the new model works with the existing API by running a small smoke test:
     - Load `xgb_model.pkl` and score 10 sample payloads.
   - Keep previous model intact; document rollback step:
     - Change `latest.txt` pointer back to prior model directory if needed.

6. **Reporting**
   - File: `docs/retraining_reports/xgb_retraining_<YYYYMMDD>.md`
   - Include:
     - Data added from feedback (counts, class mix).
     - Before/after metrics (F1, Precision, Recall, ROC-AUC, PR-AUC) on val/test.
     - Confusion matrices deltas.
     - Notes on threshold implications (if you later adjust `T_LOW`/`T_HIGH`).

7. **Make target**
   - Add to README or Makefile:
     ```bash
     python scripts/retrain_supervised.py \
       --feedback data/feedback.parquet \
       --out_dir artifacts/models/xgb
     ```

---

### Acceptance Criteria
- [ ] Feedback ingested and merged into training set without schema errors.
- [ ] Retrained XGBoost model saved with versioned path and metrics JSON.
- [ ] Validation/test metrics computed with before/after comparison table.
- [ ] `latest.txt` updated and rollback instructions documented.
- [ ] Retraining report created under `docs/retraining_reports/`.

---

### Artifacts
- `scripts/retrain_supervised.py`
- `artifacts/models/xgb/<version>/xgb_model.pkl`
- `artifacts/models/xgb/<version>/xgb_metrics.json`
- `artifacts/models/xgb/latest.txt`
- `docs/retraining_reports/xgb_retraining_<YYYYMMDD>.md`

---

### Helpful Links
- XGBoost tuning tips: https://xgboost.readthedocs.io/en/stable/tutorials/param_tuning.html
- SMOTE guidelines: https://imbalanced-learn.org/stable/references/generated/imblearn.over_sampling.SMOTE.html

---

### Effort Estimate
M (1–2 days)
